[Feature] Add `PageExportController` to bundle

mike4git opened the pull request:
By introducing this controller and other stuff we have a contextual menu for Documents:

![image](https://github.com/user-attachments/assets/743b0e13-e2ab-49d9-9277-9573c60a17f8)

jdreesen commented on config/pimcore/routing.yml:
```diff
@@ -0,0 +1,5 @@
+page_export:
+    resource: '@NeustaTrinityPimcoreBundle/Controller/Admin/PageExportController.php'
```
Wrong bundle name.

jdreesen commented on public/js/pimcore/exportPage.js:
```diff
@@ -0,0 +1,57 @@
+// /Resources/public/js/exportPage.js
+
```
Useless comment (with wrong path).

```suggestion
```

jdreesen commented on public/js/pimcore/exportPage.js:
```diff
@@ -0,0 +1,57 @@
+    });
+
+    var pimcorePluginPageExport = new pimcore.plugin.page.export();
+} else {
+    pimcore.registerNS("pimcore.plugin.page_export");
+
+    pimcore.plugin.TrinityPimcoreBundle = Class.create(pimcore.plugin.admin, {
+        getClassName: function () {
+            return "pimcore.plugin.page_export";
+        },
+
+        initialize: function () {
+            pimcore.plugin.broker.registerPlugin(this);
+        },
+
+        prepareDocumentTreeContextMenu: function (menu, tree, record) {
+            // Export page into yaml file
+            menu.add("-");
+            menu.add(new Ext.menu.Item({
+                text: 'Export to yaml',
+                iconCls: "pimcore_icon_export",
+                handler: function () {
+                    pimcore.helpers.download(Routing.generate('page_export', {page_id: record.data.id}));
+                }
+            }));
+        },
+
+    });
+
+    var trinityPimcoreBundlePlugin = new pimcore.plugin.TrinityPimcoreBundle();
+}
```
This bundle only supports Pimcore 11, so we don't need this.

jdreesen commented on public/js/pimcore/exportPage.js:
```diff
@@ -0,0 +1,57 @@
+const pimcoreVersion = parseInt(pimcore.settings.version.match(/^v(\d+)/)[1]);
+
+if (pimcoreVersion >= 11) {
+    pimcore.registerNS("pimcore.plugin.page.export");
```
I think we should use a bundle prefix here.

jdreesen commented on public/js/pimcore/exportPage.js:
```diff
@@ -0,0 +1,57 @@
+            // Export page into yaml file
+            menu.add("-");
+            menu.add(new Ext.menu.Item({
+                text: 'Export to yaml',
```
Translate?

lukadschaak replied:
Nö, erstmal nich

jdreesen commented on src/Controller/Admin/PageExportController.php:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    /**
+     * @Route("/admin/page/export", name="page_export", methods={"GET"})
```
We should prefix this with the bundle prefix:

```suggestion
     * @Route("/admin/neusta/import-export/page/export", name="neusta_pimcore_import_export_page_export", methods={"GET"})
```

jdreesen commented on src/Controller/Admin/PageExportController.php:
```diff
@@ -0,0 +1,59 @@
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+
+final class PageExportController extends UserAwareController
```
I don't think we need to extend the `UserAwareController` here :thinking:

jdreesen commented on src/EventListener/PimcoreAdminListener.php:
```diff
@@ -0,0 +1,63 @@
+
+use Pimcore\Event\BundleManager\PathsEvent;
+
+class PimcoreAdminListener
```
```suggestion
final class PimcoreAdminListener
```

jdreesen commented on src/EventListener/PimcoreAdminListener.php:
```diff
@@ -0,0 +1,63 @@
+            ],
+        ));
+    }
+
+    /**
+     * Overloads a bundle file with app file if exists.
+     */
+    private function overloadFile(string $bundleFile, string $appFile): string
+    {
+        $bundleFile = \sprintf('/bundles/neustatrinitypimcore/%s', $bundleFile);
+        $appFile = \sprintf('/neustatrinitypimcore/%s', $appFile);
+
+        return file_exists(PIMCORE_WEB_ROOT . $appFile) ? $appFile : $bundleFile;
+    }
```
Unused after above methods got removed.

jdreesen commented on src/EventListener/PimcoreAdminListener.php:
```diff
@@ -0,0 +1,63 @@
+
+    public function addJSFiles(PathsEvent $event): void
+    {
+        $event->setPaths(array_merge(
+            $event->getPaths(),
+            [
+                '/bundles/neustatrinitypimcore/js/startup.js',
+                '/bundles/neustatrinitypimcore/js/exportPage.js',
+            ],
+        ));
```
```suggestion
        $event->addPaths([
            '/bundles/neustapimcoreimportexport/js/exportPage.js',
        ]);
```

mike4git replied:
I've made a lot more changes to the Listener because most of the things we do not need and the directory naming was incorrect, too.

jdreesen replied:
But that doesn't mean we don't want this change, too.

mike4git commented on config/pimcore/routing.yml:
```diff
@@ -0,0 +1,5 @@
+page_export:
```
???

jdreesen replied:
It should be: `neusta_pimcore_import_export_page_export`

jdreesen commented:
Please also take a look at the hidden annotations:

![image](https://github.com/user-attachments/assets/6398a800-a393-47f6-8c7f-f5a093c2ef11)

mike4git commented:
I've added default bundle prefix where possible/necessary and I've added dependencies from internal bundles which are necessary for our implementation: `Neusta\Pimcore\ImportExportBundle\PimcoreConverter\Populator\PropertyBasedMappingPopulator` incl. all tests and fixture objects

jdreesen commented on src/EventListener/PimcoreAdminListener.php:
```diff
@@ -0,0 +1,18 @@
+{
+    public function addJSFiles(PathsEvent $event): void
+    {
+        $event->setPaths(array_merge(
+            $event->getPaths(),
+            [
+                '/bundles/neustapimcoreimportexport/js/exportPage.js',
+            ],
+        ));
```
```suggestion
        $event->addPaths([
            '/bundles/neustapimcoreimportexport/js/exportPage.js',
        ]);
```

jdreesen commented on src/PimcoreConverter/Context/ContextWithLocale.php:
```diff
@@ -0,0 +1,10 @@
+
+namespace Neusta\Pimcore\ImportExportBundle\PimcoreConverter\Context;
+
+interface ContextWithLocale
+{
+    public function getLocale(): ?string;
+
+    public function setLocale(?string $locale): static;
+}
```
This is unused.

jdreesen commented on tests/Unit/Controller/Admin/PageExportControllerTest.php:
```diff
@@ -0,0 +1,78 @@
+
+        $response = $this->controller->exportPage($this->request);
+
+        self::assertTrue($response instanceof Response);
```
```suggestion
        self::assertInstanceOf(Response::class, $response);
```

However, this is unnecessary as the method has a return type of `Response`, so this will *always* be the case.

jdreesen commented on tests/Unit/Controller/Admin/PageExportControllerTest.php:
```diff
@@ -0,0 +1,78 @@
+
+        $response = $this->controller->exportPage($this->request);
+
+        self::assertTrue($response instanceof JsonResponse);
```
```suggestion
        self::assertInstanceOf(JsonResponse::class, $response);
```

jdreesen commented on tests/Unit/Controller/Admin/PageExportControllerTest.php:
```diff
@@ -0,0 +1,78 @@
+
+        $response = $this->controller->exportPage($this->request);
+
+        self::assertTrue($response instanceof JsonResponse);
```
```suggestion
        self::assertInstanceOf(JsonResponse::class, $response);
```

jdreesen commented on tests/Unit/PimcoreConverter/Populator/PropertyBasedMappingPopulatorTest.php:
```diff
@@ -0,0 +1,48 @@
+{
+    use ProphecyTrait;
+
+    private ObjectProphecy|TestDataObject $target;
+    private DataObject|ObjectProphecy $source;
+    private ContextWithLocale|ObjectProphecy $context;
```
```suggestion
    /** @var ObjectProphecy<TestDataObject> */
    private ObjectProphecy $target;
    /** @var ObjectProphecy<DataObject> */
    private ObjectProphecy $source;
    /** @var ObjectProphecy<ContextWithLocale> */
    private ObjectProphecy $context;
```

jdreesen commented on tests/Unit/PimcoreConverter/Populator/PropertyBasedMappingPopulatorTest.php:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /** @test */
+    public function populateShouldPopulateTargetWithDataFromSourceWithoutLanguage(): void
+    {
+        $this->source->getProperty('language')->willReturn('de')->shouldBeCalled();
+        $this->populator->populate($this->target, $this->source->reveal(), $this->context->reveal());
+    }
```
I think this is missing an assertion.

jdreesen commented on src/Controller/Admin/PageExportController.php:
```diff
@@ -0,0 +1,62 @@
+        }
+
+        try {
+            $json = $this->pageExporter->toYaml($page);
```
This return YAML, not JSON, right?

```suggestion
            $yaml = $this->pageExporter->toYaml($page);
```

jdreesen commented on src/Controller/Admin/PageExportController.php:
```diff
@@ -0,0 +1,62 @@
+            return new JsonResponse($e->getMessage(), Response::HTTP_INTERNAL_SERVER_ERROR);
+        }
+
+        $response = new Response($json);
+        $response->headers->set('Content-type', 'application/json');
```
```suggestion
        $response = new Response($yaml);
        $response->headers->set('Content-type', 'application/yaml');
```
